pivoting/Pivoting_Study.md

```diff
@@ -1,30 +1,110 @@
+- ```clusterctl``` binary takes the source and target cluster kubeconfig and the provider components file as arguements. 
+- Using the provider components file the CRDs, CRs and controllers are created on the target cluster. Once the creation is done, the controllers are scaled down in the source cluster.
+- The CAPI objects and the BM deployment specific objects are created/moved in the target cluster and deleted in the source cluster. But the BMHs are not deleted from the source cluster.
+
+### Challenges Encountered 
+As of clusterctl release-02, pivoting lacks the following features which are critical in baremetal based deployments.
+
+1. Pivoting objects other than CAPI
+   As described in the 'Pivoting process', CRDs, CRs and controllers are created on the target cluster. Related core objects, such as secrets, are also moved. However, provider specific objects such as bmh and objecs related with them are not moved. 
+
+   We have enhanced the clusterctl code to handle moving of bmh and related ojbects.
+
+2. Moving BMH with status
+   Since we have a unique requirement of pivoting BMH object with status preserved from source cluster, we added this feature to patch the BMH object in the target cluster to preserve status from the source cluster. 
+
+3. Maintaining connectivity towards provisioning network
+   Baremetal machines boot over a network with a dhcp server. This requires maintaing a fixed IP end points towards the provisioning network. We achieved this using keepalived. 
+
+4. Maintaining dhcp leases 
+   During the pivoting process, the pool of baremetal physical machines remains the same. As to avoid IP address assignment conflicts,  maintaining the same dhcp leases in the target cluster is required.
+   To this end, we have added a configmap with the content of dhcp leases from the source cluster. The configmap is mounted as a volume in the dnsmasq container of the BMO controller in the target cluster. This ensured that the dnsmasq container is booted with the right leases content.
+
+### Future works
+While working with the current version ```clusterctl```, we have observed the following inconsistencies and propose the following enhancements.
+
+1. Linking objects
+   Currently, objects are linked using labels, annotations, OwnerRefernces and ConsumerReferences. We need to choose one method that makes the pivoting code generic enough.
+2. PivotedReady flag for moved objects
+   Once an object is moved to a target cluster, there is needs to be a uniform flag that indicates it has been successfully created and any subsequent objects can be created. 
+3. RollBack failed pivoting process
+   Currently, if pivoting fails, both the target and source clusters are left in an undesired state. The pivoting process needs to be atomic, i.e. all or none.
+
+## Pivoiting Steps
+Please run the following commands to perform pivoting.
+replace the target TARGET_CLUSTER with the IP of the target cluster
+
+```bash
+#!/bin/bash
```

Comment:
script file exists?
please refer to the file here instead of listing the commands

Reply:
Kept the commands for better understanding as well as visibility of whats happening in those steps. Will remove this once the whole process is finalized.